### PR TITLE
builtins/printf: avoid string copies by formatting directly to buffer

### PR DESCRIPTION
**Pull request by Xiretza**

### CODE REVIEW

File: fish-rust/src/builtins/printf.rs [hunk 1]
```diff
@@ -262,21 +262,22 @@ impl<'a, 'b> builtin_printf_state_t<'a, 'b> {
...
-                    sprintf_locale(
-                        &mut target,
-                        $fmt,
-                        &self.locale,
-                        &[$($arg.to_arg()),*]
-                    );
-                    target
+                // Don't output if we're done.
+                if self.early_exit {
+                    return;
```
Comment (krobelus) [inline fish-rust/src/builtins/printf.rs:272]: should probably not return from the surrounding function, only skip the rest of the macro

**Comment by krobelus** [comment]
good idea. In future we want the same for `OutputStream`, maybe we can define a trait

File: fish-rust/src/builtins/printf.rs (hunk 1 above)
Comment (Xiretza) [reply fish-rust/src/builtins/printf.rs:272]: Whoops, absolutely right - the danger of macros...

**Comment by Xiretza** [comment]
I think `printf_compat::output::WideWrite` is that trait, just needs to be implemented for `OutputStream`.

**Pull request closed without merging**